Blog post: Embedding JavaScript snippets in C++ (#7162)
GoogleChrome/web.dev: src/site/content/en/blog/emscripten-embedding-js-snippets/index.md

---
title: Embedding JavaScript snippets in C++ with Emscripten
subhead: Learn how to embed JavaScript code in your WebAssembly library to communicate with the outside world.
description: Learn how to embed JavaScript code in your WebAssembly library to communicate with the outside world.

> date: 2022-01-20

Author: ```suggestion
date: 2022-01-17
```

@rachelandrew If this is looking good otherwise, then I'm happy to publish today.

Author: Or today :)

```suggestion
date: 2022-01-18
```

hero: image/9oK23mr86lhFOwKaoYZ4EySNFp02/InR77BkPIakxD8X1rADr.jpg
alt: A man putting sticky notes on a whiteboard.
authors:
  - rreverser
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - webassembly
  - javascript
---

> When working on WebAssembly integration with the web, you need a way to call out to external APIs such as web APIs and third-party libraries. You then need a way to store the values and object instances those APIs return, and a way to pass those stored values to other APIs later. For asynchronous APIs, you might also need to await `Promise`s in your synchronous C/C++ code with [Asyncify](/asyncify/) and read the result once the operation is finished.

Reviewer: ```suggestion
When working on WebAssembly integration with the web, you need a way to call out to external APIs such as web APIs and third-party libraries. You then need a way to store the values and object instances those APIs return, and a way to pass those stored values to other APIs later. For asynchronous APIs, you might also need to await promises in your synchronous C/C++ code with [Asyncify](/asyncify/) and read the result once the operation is finished.
```

```cpp
#include <emscripten/val.h>

using namespace emscripten;

val fetch_json(const char *url) {
  // Get and cache a binding to the global `fetch` API in each thread.
  thread_local const val fetch = val::global("fetch");
  // Invoke fetch and await the returned `Promise<Response>`.
  val response = fetch(url).await();
  // Ask to read the response body as JSON and await the returned `Promise<any>`.
  val json = response.call<val>("json").await();
  // Return the JSON object.
  return json;
}

// Example URL.
val example_json = fetch_json("https://httpbin.org/json");

// Now we can extract fields, e.g.
std::string author = json["slideshow"]["author"].as<std::string>();
```

This code works well, but it performs lots of intermediate steps. Each operation on `val` needs to perform the following steps:

1. Convert C++ values passed as arguments into some intermediate format.
2. Go to JavaScript, read and convert arguments into JavaScript values.
3. Execute the function
4. Convert the result from JavaScript to intermediate format.
5. Return the converted result to C++, and C++ finally reads it back.

> Each `await()` also has to pause the C++ side by unwinding the entire call stack of the WebAssembly module, returning to JavaScript, waiting, and restoring the WebAssembly stack when the operation is complete. \

Reviewer: ```suggestion
Each `await()` also has to pause the C++ side by unwinding the entire call stack of the WebAssembly module, returning to JavaScript, waiting, and restoring the WebAssembly stack when the operation is complete.
```

std::string value = input["value"].as<std::string>();
```

With those APIs in mind, the `fetch_json` example could be rewritten to do most work without leaving JavaScript:

```cpp
EM_JS(EM_VAL, fetch_json, (const char *url), {
  return Asyncify.handleAsync(async () => {
    url = UTF8ToString(url);
    // Invoke fetch and await the returned `Promise<Response>`.
    let response = await fetch(url);
    // Ask to read the response body as JSON and await the returned `Promise<any>`.
    let json = await response.json();
    // Convert JSON into a handle and return it.
    return Emval.toHandle(json);
  });
});

// Example URL.
val example_json = val::take_ownership(fetch_json("https://httpbin.org/json"));

// Now we can extract fields, e.g.
std::string author = json["slideshow"]["author"].as<std::string>();
```

We still have a couple of explicit conversions at the entry and exit points of the function, but the rest is now regular JavaScript code. Unlike `val` equivalent, it can now be optimized by the JavaScript engine and only requires pausing the C++ side once for all async operations.

## EM_ASYNC_JS macro

> The only bit left that does not look pretty is the `Asyncify.handleAsync` wrapper - its only purpose is to allow executing `async` JavaScript functions with Asyncify. In fact, this use case is so common that there is now a specialized `EM_ASYNC_JS` macro that combines them together.

Reviewer: ```suggestion
The only bit left that does not look pretty is the `Asyncify.handleAsync` wrapper—its only purpose is to allow executing `async` JavaScript functions with Asyncify. In fact, this use case is so common that there is now a specialized `EM_ASYNC_JS` macro that combines them together.
```

In some cases, however, you want to insert a quick snippet for [`console.log`](https://developer.mozilla.org/docs/Web/API/Console/log) call, a [`debugger;`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/debugger) statement or something similar and don't want to bother with declaring a whole separate function. In those rare cases, an [`EM_ASM macros family`](https://emscripten.org/docs/api_reference/emscripten.h.html#c.EM_ASM) (`EM_ASM`, `EM_ASM_INT` and `EM_ASM_DOUBLE`) might be a simpler choice. Those macros are similar to the `EM_JS` macro, but they execute code inline where they're inserted, instead of defining a function.

Since they don't declare a function prototype, they need a different way of specifying the return type and accessing arguments.

You need to use the right macro name to choose the return type. `EM_ASM` blocks are expected to act like `void` functions, `EM_ASM_INT` blocks can return an integer value, and `EM_ASM_DOUBLE` blocks return floating-point numbers correspondingly.

> Any passed arguments will be available under names `$0`, `$1`, and so on in the JavaScript body. As with `EM_JS` or WebAssembly in general, the arguments are limited only to numeric values - integers, floating-point numbers, pointers and handles.

Reviewer: ```suggestion
Any passed arguments will be available under names `$0`, `$1`, and so on in the JavaScript body. As with `EM_JS` or WebAssembly in general, the arguments are limited only to numeric values—integers, floating-point numbers, pointers and handles.
```
